### DISCUSSION TITLE
Using build to build a PDM project

### DISCUSSION THREAD

**Discussion by pawamoy** [pdm-project/pdm#440] [Q&A] [answered]
https://pypa-build.readthedocs.io/en/latest/index.html

Is it possible? Here is some output that shows it doesn't seem to work:

```console
$ pyproject-build src -o dist
running egg_info
writing UNKNOWN.egg-info/PKG-INFO
writing dependency_links to UNKNOWN.egg-info/dependency_links.txt
writing top-level names to UNKNOWN.egg-info/top_level.txt
reading manifest file 'UNKNOWN.egg-info/SOURCES.txt'
writing manifest file 'UNKNOWN.egg-info/SOURCES.txt'
running sdist
running egg_info
writing UNKNOWN.egg-info/PKG-INFO
writing dependency_links to UNKNOWN.egg-info/dependency_links.txt
writing top-level names to UNKNOWN.egg-info/top_level.txt
reading manifest file 'UNKNOWN.egg-info/SOURCES.txt'
writing manifest file 'UNKNOWN.egg-info/SOURCES.txt'
warning: sdist: standard file not found: should have one of README, README.rst, README.txt, README.md

running check
warning: check: missing required meta-data: name, url

warning: check: missing meta-data: either (author and author_email) or (maintainer and maintainer_email) must be supplied

creating UNKNOWN-0.0.0
creating UNKNOWN-0.0.0/UNKNOWN.egg-info
copying files to UNKNOWN-0.0.0...
copying UNKNOWN.egg-info/PKG-INFO -> UNKNOWN-0.0.0/UNKNOWN.egg-info
copying UNKNOWN.egg-info/SOURCES.txt -> UNKNOWN-0.0.0/UNKNOWN.egg-info
copying UNKNOWN.egg-info/dependency_links.txt -> UNKNOWN-0.0.0/UNKNOWN.egg-info
copying UNKNOWN.egg-info/top_level.txt -> UNKNOWN-0.0.0/UNKNOWN.egg-info
Writing UNKNOWN-0.0.0/setup.cfg
Creating tar archive
removing 'UNKNOWN-0.0.0' (and everything under it)


running egg_info
writing UNKNOWN.egg-info/PKG-INFO
writing dependency_links to UNKNOWN.egg-info/dependency_links.txt
writing top-level names to UNKNOWN.egg-info/top_level.txt
reading manifest file 'UNKNOWN.egg-info/SOURCES.txt'
writing manifest file 'UNKNOWN.egg-info/SOURCES.txt'
running bdist_wheel
running build
installing to build/bdist.linux-x86_64/wheel
running install
running install_egg_info
running egg_info
writing UNKNOWN.egg-info/PKG-INFO
writing dependency_links to UNKNOWN.egg-info/dependency_links.txt
writing top-level names to UNKNOWN.egg-info/top_level.txt
reading manifest file 'UNKNOWN.egg-info/SOURCES.txt'
writing manifest file 'UNKNOWN.egg-info/SOURCES.txt'
Copying UNKNOWN.egg-info to build/bdist.linux-x86_64/wheel/UNKNOWN-0.0.0-py3.8.egg-info
running install_scripts
creating build/bdist.linux-x86_64/wheel/UNKNOWN-0.0.0.dist-info/WHEEL
creating 'dist/tmpldoj41te/UNKNOWN-0.0.0-py3-none-any.whl' and adding 'build/bdist.linux-x86_64/wheel' to it
adding 'UNKNOWN-0.0.0.dist-info/METADATA'
adding 'UNKNOWN-0.0.0.dist-info/WHEEL'
adding 'UNKNOWN-0.0.0.dist-info/top_level.txt'
adding 'UNKNOWN-0.0.0.dist-info/RECORD'
removing build/bdist.linux-x86_64/wheel
```

**Comment by pawamoy** [comment]
Ah, my bad, the positional argument of `pyproject-build` is supposed to be the directory where `pyproject.toml` can be found.

**Reply by pawamoy** [reply]
However now it seems that no source is added at all, while tests, README.md and pyproject.toml are added:

```console
$ pyproject-build
# sdist
 - Adding README.md
 - Adding pyproject.toml
 - Adding tests/__init__.py
 - Adding tests/conftest.py
 - Adding tests/test_auth.py
 - Adding tests/test_sson.py
 - Adding PKG-INFO

# wheel
 - Adding pkg_name-0.4.3.dev7+gd868f34.d20210504.dist-info/entry_points.txt
 - Adding pkg_name-0.4.3.dev7+gd868f34.d20210504.dist-info/WHEEL
 - Adding pkg_name-0.4.3.dev7+gd868f34.d20210504.dist-info/METADATA
 - Adding pkg_name-0.4.3.dev7+gd868f34.d20210504.dist-info/RECORD
```

**Reply by pawamoy** [reply]
Ah, I think the PEP517 backend does not properly detect namespace packages!

My package resides in `src/namespace/package`:

```
root/
    src/
        namespace/
            package/
                __init__.py
```

**Reply by pawamoy** [reply]
In short: how should I configure PDM to retrieve a namespace package from `src`? Reading the docs, I'm not sure if I should use `includes` or `package-dir` or both or none.

**Reply by pawamoy** [reply]
OK it's working if I set this:

```toml
[tool.pdm]
includes = ["src/namespace"]
```

I thought that the following would suffice, but it seems not:

```toml
[tool.pdm]
package-dir= "src"
```

**Comment by pawamoy** [comment, marked as answer]
Definitely possible!
The `pyproject-build` tool expects the directory containing `pyproject.toml` as a positional argument. It defaults to `.` so you can instead run it without any arguments:

```
pyproject-build
```

Next, for namespace packages (and it has nothing to do with `build` but rather PDM and `pdm.pep517`), I had to explicitely include the namespace package with 

```toml
[tool.pdm]
includes = ["src/namespace"]
```

because this was not enough:

```toml
[tool.pdm]
package-dir= "src"
```

**Reply by frostming** [reply]
No, we don't auto detect **implicit** namespace packages otherwise lots of chore directories will be in luded without awareness. In setuptools you also needs find_namespace_packages yeah? So it is reasonable to require explicit includes i think.

**Reply by pawamoy** [reply]
Yes I think setuptools also required either explicit listing or the `find_namespace` thing 🙂 
Then I think [point 3](https://pdm.fming.dev/pyproject/tool-pdm/#implicit-namespace-packages) should be clarified, because I thought it meant that PDM would not detect the implicit namespace package if I used `package-dir = "src/namespace"`, while I was using `package-dir "src"`. Maybe something like this:

> 3. `<package>` is not contained in the specified `package-dir`

or

> 3. `package-dir` is not used

or 

> 3. `<package>` is explicitly listed in `includes`

**Reply by frostming** [reply]
That might be outdated. In fact there isn't any autodection for namespace packages

**Reply by frostming** [reply]
The reason behind this is it's hard to distinguish the following directories

- Plain directories
- namespace packages
- src directory that should be detected as package-dir

**Reply by pawamoy** [reply]
I understand. A simple update to the docs would be perfect 🙂 If it's only a matter of removing that paragraph, I can send the PR :smile:
